### PR TITLE
Bundle with tsup and replace babel-jest

### PR DESCRIPTION
**Pull request by bryanjtc**
Bundles code and some dependencies with tsup.
Replaces babel-jest with @swc/core. With swc tests should run faster than using babel.
Adds require.resolve for jest-junit and jest-serializer-html.
Adds support for esm only projects that don't use a custom config file.
Tested locally using test-storybook:ci commands and some projects that I have with storybook and test-runner.

<details>
  <summary>📦 Published PR as canary version: <code>1.0.0--canary.224.3c57b66.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @storybook/test-runner@1.0.0--canary.224.3c57b66.0
  # or 
  yarn add @storybook/test-runner@1.0.0--canary.224.3c57b66.0
  ```
</details>

### CODE REVIEW

**Comment by bryanjtc** [comment]
@yannbf Can you check this?

**Comment by bryanjtc** [comment]
@yannbf What do you think of these changes?

**Comment by yannbf** [comment]
Hey @bryanjtc thanks for your contribution! Sorry, thing have been busy with SB beta. We'll take a look!

File: package.json [hunk 1]
```diff
@@ -16,35 +16,27 @@
...
   "license": "MIT",
-  "main": "dist/cjs/index.js",
-  "module": "dist/esm/index.js",
-  "types": "dist/ts/index.d.ts",
+  "main": "dist/index.js",
+  "module": "dist/index.mjs",
+  "types": "dist/index.d.ts",
   "bin": {
-    "test-storybook": "./bin/test-storybook.js"
+    "test-storybook": "./dist/test-storybook"
```
Comment (ndelangen) [inline package.json:23]: Is it OK to have no file-extension here?

File: src/playwright/transformPlaywrightJson.ts [hunk 2]
```diff
@@ -86,7 +86,7 @@ export const transformPlaywrightJson = (index: Record<string, any>) => {
     if (stories.length) {
       const storyTests = stories.map((story) => makeDescribe(story.name, [makeTest(story)]));
       const program = t.program([makeDescribe(stories[0].title, storyTests)]);
-
+      // @ts-ignore
```
Comment (ndelangen) [inline src/playwright/transformPlaywrightJson.ts:89]: I think this should have a reason added:

```suggestion
      // @ts-ignore (why is this OK? who should fix this / when?)
```

Comment (ndelangen) [reply src/playwright/transformPlaywrightJson.ts:89]: Don't actually commit ☝️, should have the actual reason (which I do not know)

File: tsconfig.json [hunk 3]
```diff
@@ -13,7 +13,7 @@
     "noImplicitAny": true,
     "rootDir": "./src",
     "skipLibCheck": true,
-    "target": "es5",
+    "target": "es6",
```
Comment (ndelangen) [inline tsconfig.json:16]:
```suggestion
    "target": "es2020",
```

**Comment by ndelangen** [comment]
Just some finer questions/points really. Looks really good, in fact, @bryanjtc I'd love to get that `@swc/jest` change in the storybook monorepo some day too.

FYI: I have a chain or PRs on there removing babel: (all except for unit test 👎 ) (and code-mods 👍 )
https://github.com/storybookjs/storybook/pull/20156

File: tsup.config.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    entry: ['./src/**/!(*.{d,test}).{js,jsx,ts,tsx}'],
+    format: ['cjs', 'esm'],
+    dts: true,
+    esbuildOptions(options, context) {
+      options.platform = 'node';
+    },
+  },
+  {
+    clean: true,
+    entry: ['bin'],
```
Comment (ndelangen) [inline tsup.config.ts:15]: We should just move the file into `src` then?

File: package.json (hunk 1 above)
Comment (bryanjtc) [reply package.json:23]: Yes. Node automatically uses test-storybook.js for cjs projects and test-storybook.mjs for esm projects.

File: tsup.config.ts (hunk 4 above)
Comment (bryanjtc) [reply tsup.config.ts]: Done

File: tsconfig.json (hunk 3 above)
Comment (bryanjtc) [reply tsconfig.json]: Done

File: src/playwright/transformPlaywrightJson.ts (hunk 2 above)
Comment (bryanjtc) [reply src/playwright/transformPlaywrightJson.ts]: I got rid of almost all ts-ignore comments. and fixed the errors.

**Comment by ndelangen** [comment]
@yannbf Is there no CI set up for this project?

**Review by ndelangen — APPROVED** [review]

File: src/playwright/transformPlaywrightJson.ts (hunk 2 above)
Comment (yannbf) [reply src/playwright/transformPlaywrightJson.ts]: Probably `ts-expect-error` instead

**Comment by yannbf** [comment]
> @yannbf Is there no CI set up for this project?

There is, I guess it doesn't run on forks? I'll take a look at this tomorrow

**Comment by bryanjtc** [comment]
> > @yannbf Is there no CI set up for this project?
> 
> There is, I guess it doesn't run on forks? I'll take a look at this tomorrow

The pr needs a label for the ci to run. Can you add it?

**Comment by ndelangen** [comment]
I don't know what labels to apply to make the right things happen 😞

**Comment by bryanjtc** [comment]
@yannbf Any update on getting ci to work?

**Comment by ndelangen** [comment]
@yannbf is this supposed to work:
```
[SB] ERR! Module not found: Error: Can't resolve './stories/stress-test' in '/home/runner/work/test-runner/test-runner'
[SB] ERR! ModuleNotFoundError: Module not found: Error: Can't resolve './stories/stress-test' in '/home/runner/work/test-runner/test-runner'
```

**Comment by bryanjtc** [comment]
I enabled some ci workflows for pull requests and modified them to run in node 16 and 18. http-server doesn't work with node 18, so I replaced it with serve. Snapshots were updated, nothing major, it just removed some empty lines.

**Comment by yannbf** [comment]
Hey @bryanjtc thank you for all you work in this! I am currently on holidays, so I'll be checking this thoroughly once I'm back next week, I hope you don't mind!

**Comment by yannbf** [comment]
Hey @bryanjtc I took a quick look, and to make a proper canary release I made another branch off of this one, but somehow that made this fork be able to release as well, but given this PR had "major" in it, it made a 1.0.0 version 🤦

With a canary available (and subsequent, as you make new pushes to this branch), let's test it out in our projects.

When testing, the binary did not exist anymore, so I got `Command "test-storybook" not found` when running `yarn test-storybook` in a project which was previously working.
I like what you did to the `test-storybook.js` file, but unfortunately it's not so easily understood when inspecting it via node_modules and easily modifiable so people could try quick fixes. But I guess that's expected if we want to support cjs/esm. Maybe if the prebundled code came from a vendors file, it would still be relatively simple to understand 🤔

**Comment by bryanjtc** [comment]
Node should be able to find the bin file without extension. By default it should use `test-storybook.js` and for ESM only projects `test-storybook.mjs`. When specifying the extension, ESM projects expect `test-storybook.js` to be esm code. There are two solutions for this, changing the extension to `.cjs` or not specifying the extension. I can undo the prebundling for the dependencies of test-storybook.ts, so that it can be easier to read and modify in `node_modules`.

**Comment by ndelangen** [comment]
@yannbf let's discuss this Monday?

**Comment by bryanjtc** [comment]
Any update on this?

**Comment by bryanjtc** [comment]
Any issues trying to re-create the pr? @yannbf

**Comment by JReinhold** [comment]
This looks great!

@ndelangen will add you to the Storybook organization, and @yannbf will redo the PR to fix the CI issue.

**Comment by ndelangen** [comment]
I'll re-open this PR

**Comment by ndelangen** [comment]
done:
https://github.com/storybookjs/test-runner/pull/250

**Pull request closed without merging**